usr/lib/common/pkey_utils.c

```diff
@@ -161,6 +161,42 @@ static int s390_kmc(unsigned long func, void *param, unsigned char *dest,
     return func ? src_len - __src_len : __src_len;
 }
 
+/**
+ * s390_kdsa:
+ * @func: the function code passed to KDSA; see s390_kdsa_functions
+ * @param: address of parameter block; see POP for details on each func
+ * @src: address of source memory area
+ * @srclen: length of src operand in bytes
+ *
+ * Executes the KDSA (COMPUTE DIGITAL SIGNATURE AUTHENTICATION) operation of
+ * the CPU.
+ *
+ * Returns 0 on success. Fails in case of sign if the random number was not
+ * invertible. Fails in case of verify if the signature is invalid or the
+ * public key is not on the curve.
+ */
+static int s390_kdsa(unsigned long func, void *param,
+                    const unsigned char *src, unsigned long srclen)
+{
+    register unsigned long r0 __asm__("0") = (unsigned long)func;
+    register unsigned long r1 __asm__("1") = (unsigned long)param;
+    register unsigned long r2 __asm__("2") = (unsigned long)src;
+    register unsigned long r3 __asm__("3") = (unsigned long)srclen;
+    unsigned long rc = 1;
+
+    __asm__ volatile(
+        "0: .insn   rre,%[__opc] << 16,0,%[__src]\n"
+        "   brc 1,0b\n" /* handle partial completion */
+        "   brc 7,1f\n"
```

Comment:
Not sure, but wouldn't `brc 6,1f\n` be sufficient here?  Due to the line before, `cc=3` (mask value position 1) is already covered and cannot be set here anymore.